### PR TITLE
Refactored serialize.py to allow for de-serialization of normalized reward functions

### PR DESCRIPTION
**Pull request by levmckinney**
Within `serialize.py`, I've broke down `_load_reward_net_as_fn` into a series of composable functions that should allow for more flexibility in loading reward functions in the future. I have maintained the public interface of `load_reward` to avoid a breaking change. Finally, I've added a new reward function type that can be de-serialized `RewarNet_normalized` and `RewardNet_unnormalized`.  The former allows the user to maintain the normalization learned during training while the latter strips off the normalization wrapper and returns the underlying unnormalized reward function. In both cases the reward function will stop updating its normalization statistics after it is de-serialized.

### CODE REVIEW

**Comment by levmckinney** [comment]
This should resolve #437 since now train_rl can be called with the normalize flag set to false and the reward type set to `RewardNet_normalized`.

File: src/imitation/rewards/serialize.py [hunk 1]
```diff
@@ -62,14 +100,36 @@ def f(
 
 
 # TODO(adam): I think we can get rid of this and have just one RewardNet.
+
 reward_registry.register(
     key="RewardNet_shaped",
-    value=_load_reward_net_as_fn(shaped=True),
+    value=lambda path, _: _validate_reward(_make_functional(th.load(str(path)))),
 )
 reward_registry.register(
```
Comment (yawen-d) [inline src/imitation/rewards/serialize.py:108]: Seems to me `NormalizedRewardNet` and `ShapedRewardNet` are both subclasses of `RewardNetWrapper`. Could simplify things by registering a `"RewardNet_unwrapped"`.

File: src/imitation/rewards/serialize.py [hunk 2]
```diff
@@ -62,14 +100,36 @@ def f(
...
+    value=lambda path, _: _validate_reward(
+        _make_functional(
+            _validate_type(th.load(str(path)), NormalizedRewardNet),
+            attr="predict_processed",
+            kwargs={"update_stats": False},
+        ),
+    ),
+)
+
+reward_registry.register(
```
Comment (yawen-d) [inline src/imitation/rewards/serialize.py:126]: Same comments as previous ones

File: src/imitation/rewards/serialize.py [hunk 3]
```diff
@@ -1,12 +1,18 @@
...
 
-from typing import Callable
+from typing import Callable, List, Type
 
 import numpy as np
 import torch as th
 from stable_baselines3.common.vec_env import VecEnv
 
 from imitation.rewards import common
+from imitation.rewards.reward_nets import (
```
Comment (yawen-d) [inline src/imitation/rewards/serialize.py:10]: Could simply this by only importing `RewardNetWrapper`.

File: src/imitation/rewards/serialize.py [hunk 4]
```diff
@@ -17,33 +23,65 @@
...
+
+
+def _strip_wrappers(
+    net: RewardNet,
+    wrappers: List[Type[RewardNetWrapper]],
+) -> RewardNet:
+    """Attempts to remove provided wrappers.
+
+    wrappers until either it reaches a non-warper
+    reward net or it has removed all the listed wrappers.
```
Comment (yawen-d) [inline src/imitation/rewards/serialize.py:49]: Strip wrappers until either it reaches a non-warpped reward net or it has removed all the listed wrappers.

File: src/imitation/rewards/serialize.py [hunk 5]
```diff
@@ -17,33 +23,65 @@
...
+
+    Args:
+        net: an instance of a reward network that may be wrapped
+        wrappers: list of wrapper types to remove
+
+    Returns:
+        The reward network with the listed wrappers removed
+    """
+    for wrapper_type in wrappers:
```
Comment (yawen-d) [inline src/imitation/rewards/serialize.py:58]: I would suggest simplying this part using a while loop, similar to [here](https://github.com/HumanCompatibleAI/imitation/blob/f332680115380ce654cfd061fbaaa7bc47918d2c/src/imitation/algorithms/adversarial/airl.py#L98).
```python
# Recursively return the base network of the wrapped reward net
while isinstance(reward_net, reward_nets.RewardNetWrapper):
    reward_net = reward_net.base
```

Though it would be better to add more detailed comments here. WDYT?

File: src/imitation/rewards/serialize.py [hunk 6]
```diff
@@ -17,33 +23,65 @@
...
+        if not isinstance(net, RewardNetWrapper):
+            # Reached base reward net
+            break
+
+        assert hasattr(net, "base")
```
Comment (yawen-d) [inline src/imitation/rewards/serialize.py:63]: `base` is a property of `RewardNetWrapper`. I think no need to check here.

File: src/imitation/rewards/serialize.py [hunk 7]
```diff
@@ -17,33 +23,65 @@
...
-    return loader
+def _make_functional(
+    net: RewardNet,
+    attr: str = "predict",
+    kwargs=dict(),
+) -> common.RewardFn:
+    return lambda *args: getattr(net, attr)(*args, **kwargs)
+
+
+def _validate_type(net: RewardNet, type_: Type[RewardNet]):
```
Comment (yawen-d) [inline src/imitation/rewards/serialize.py:81]: Change the function name to sth like `_validate_norm_reward_type`? Because the check doesn't apply to `RewardNet_shaped`. Another alternative would be to also extend it to `RewardNet_shaped`?

File: src/imitation/scripts/train_rl.py [hunk 8]
```diff
@@ -102,6 +105,10 @@ def train_rl(
         # so normalizing it makes training more stable. Note we do *not* normalize
         # observations here; use the `NormalizeFeaturesExtractor` instead.
         venv = VecNormalize(venv, norm_obs=False, **normalize_kwargs)
+        if reward_type == "RewardNet_normalized":
```
Comment (yawen-d) [inline src/imitation/scripts/train_rl.py:108]: Maybe we should ban the behavior of double normalization. This could produce silent error if users missed the warning. I would set an Assertion error check here.

File: src/imitation/scripts/train_rl.py [hunk 9]
```diff
@@ -53,7 +54,9 @@ def train_rl(
 
     Args:
         total_timesteps: Number of training timesteps in `model.learn()`.
-        normalize_reward: If True, then rescale and clip reward.
+        normalize_reward: Applies normalization and clipping to the reward function by
```
Comment (yawen-d) [inline src/imitation/scripts/train_rl.py:57]: Good point! Thanks for making a note about the difference between `VecNormalize` and `RunningNorm`. Though the difference is minor, I would suggest making an issue to track this difference and considering wrapping the `VecEnv` using a `RewardVecEnvWrapper` that takes in a `reward_fn` that takes in the ground truth reward and pass through a normalization layer.

This would be also useful if we want to use EMWA as the normalization layer for `train_rl`.

File: tests/test_scripts.py [hunk 10]
```diff
@@ -366,6 +369,31 @@ def test_transfer_learning(tmpdir: str) -> None:
     _check_rollout_stats(run.result)
 
 
+def test_rl_train_warning(tmpdir: str):
```
Comment (yawen-d) [inline tests/test_scripts.py:372]: Depending on the changes in `train_rl.py`.

**Comment by yawen-d** [comment]
Sorry for getting to this a bit late. Thanks for the PR. I left a few comments and suggestions:
- Some code simplification in `reward_nets.py`
- Might be good to make an issue to track difference between `RunningNorm` and `VecNormalize`.

File: src/imitation/rewards/serialize.py (hunk 5 above)
Comment (levmckinney) [reply src/imitation/rewards/serialize.py:58]: I guess I was imagining the case where we have a reward network that is shaped and then normalized. Maybe the user just wants to remove the normalization but not the shaping.

File: src/imitation/rewards/serialize.py (hunk 7 above)
Comment (levmckinney) [reply src/imitation/rewards/serialize.py:81]: Ya I could definitely have it also validate that the reward function you are loading is shaped. I was just a bit nervous about breaking the current interface for anyone already using the code.

File: src/imitation/scripts/train_rl.py (hunk 8 above)
Comment (levmckinney) [reply src/imitation/scripts/train_rl.py:108]: Sure I can make it an assertion error.

File: src/imitation/rewards/serialize.py (hunk 5 above)
Comment (levmckinney) [reply src/imitation/rewards/serialize.py:58]: See https://github.com/HumanCompatibleAI/imitation/issues/437#issuecomment-1136575041

File: src/imitation/scripts/train_rl.py (hunk 9 above)
Comment (levmckinney) [reply src/imitation/scripts/train_rl.py:56]: We could do that for consistency, but it seems a bit beyond the scope of what we are doing in this pull request. I've opened another issue to discuss it #442 like you suggested.

File: src/imitation/rewards/serialize.py (hunk 1 above)
Comment (levmckinney) [reply src/imitation/rewards/serialize.py:109]: How should "RewardNet_unwrapped" behave in the presence of multiple wrappers? See https://github.com/HumanCompatibleAI/imitation/issues/437#issuecomment-1136575041.

File: src/imitation/rewards/serialize.py (hunk 5 above)
Comment (yawen-d) [reply src/imitation/rewards/serialize.py:58]: That makes sense. Thanks for reminding me of that.

File: tests/test_scripts.py [hunk 11]
```diff
@@ -358,14 +361,39 @@ def test_transfer_learning(tmpdir: str) -> None:
         named_configs=["cartpole"] + ALGO_FAST_CONFIGS["rl"],
         config_updates=dict(
             common=dict(log_dir=log_dir_data),
-            reward_type="RewardNet_shaped",
+            reward_type="RewardNet_unshaped",
```
Comment (levmckinney) [inline tests/test_scripts.py:364]: This started throwing errors once I added validation to ensure that reward functions where actually shaped when they were loaded as shaped. I think this is supposed to be unshaped since it's loading the test reward function. Is that correct? Or is this supposed to be a shaped reward?

File: src/imitation/rewards/serialize.py (hunk 7 above)
Comment (levmckinney) [reply src/imitation/rewards/serialize.py:79]: It turns out adding validation did break some test cases. I think there is likely either some caused by some other bug in the code or these test cases where not correct. See: https://github.com/HumanCompatibleAI/imitation/pull/440#discussion_r885117806

File: experiments/transfer_learn_benchmark.sh [hunk 12]
```diff
@@ -87,7 +87,7 @@ parallel -j 25% --header : --results "${LOG_ROOT}/parallel/" --colsep , --progre
   with \
   '{env_config_name}' seed='{seed}' \
   common.log_dir="${LOG_ROOT}/${ALGORITHM}/{env_config_name}_{seed}/n_expert_demos_{n_expert_demos}" \
-  reward_type="RewardNet_shaped" \
+  reward_type="RewardNet_unshaped" \
```
Comment (levmckinney) [inline experiments/transfer_learn_benchmark.sh:90]: Same issue as the previous one comment.

File: tests/test_scripts.py (hunk 11 above)
Comment (AdamGleave) [reply tests/test_scripts.py:363]: I think this is right, good catch!

**Review by AdamGleave — CHANGES_REQUESTED** [review]
Thanks for the PR! Overall looks like a good implementation, I've got a few nIt-picky changes requested.

File: src/imitation/scripts/train_rl.py [hunk 13]
```diff
@@ -80,6 +82,10 @@ def train_rl(
 
     Returns:
         The return value of `rollout_stats()` using the final policy.
+
+    Raises:
```
[the remark this answers is not in the document]
Comment (yawen-d) [reply src/imitation/scripts/train_rl.py:86]: In general, on the one hand, I think it's dubious to use VecNormalize. On the other hand, I think we should track the reward fed into the RL algorithm.

For not using `VecNormalize`:
- Let's say the transfer distribution for learned reward shifted a lot from a scale of mean 0 to 100. It might be better to shift the mean back to 0 rather than only rescaling its variance using VecNormalize. We might also want to do it for train_rl using ground truth reward, otherwise the reward scale will be naturally different between GT reward and learned reward (normalized from train norm stats).
- For a reward with arbitrary scale of mean, we are not sure how to set the `clip_reward` scale.

My current best guess are to
1. Keep the norm stats updated during retraining/transfer.
2. Or wrap the loaded reward model with ... another `RewardWrapper` where we can easily choose which norm we wanna use between RunningNorm and EWMA.

At the same time, I think it's okay to keep the VecNormalize option on until we decide whether to replace VecNormalize.

[the remark this answers is not in the document]
Comment (levmckinney) [reply src/imitation/scripts/train_rl.py:86]: Yawen and I had a discussion about having this be an error or a warning https://github.com/HumanCompatibleAI/imitation/pull/440#discussion_r885062805. I can change it back to being a warning fairly easily.

File: src/imitation/rewards/serialize.py [hunk 14]
```diff
@@ -17,33 +23,63 @@
...
+
+
+def _strip_wrappers(
+    net: RewardNet,
+    wrappers: List[Type[RewardNetWrapper]],
+) -> RewardNet:
+    """Attempts to remove provided wrappers.
+
+    Strips wrappers until either it reaches a non-warper
+    reward net or it has removed all the listed wrappers.
```
[the remark this answers is not in the document]
Comment (levmckinney) [reply src/imitation/rewards/serialize.py]: I updated description to make this more clear.

File: src/imitation/rewards/serialize.py [hunk 15]
```diff
@@ -17,33 +23,63 @@
...
+
+    Args:
+        net: an instance of a reward network that may be wrapped
+        wrappers: list of wrapper types to remove
+
+    Returns:
+        The reward network with the listed wrappers removed
+    """
+    for wrapper_type in wrappers:
+        if not isinstance(net, RewardNetWrapper):
```
[the remark this answers is not in the document]
Comment (levmckinney) [reply src/imitation/rewards/serialize.py]: Changed this into an assert to make sure the input has type `RewardNetWrapper`.

File: tests/rewards/test_reward_nets.py [hunk 16]
```diff
@@ -16,7 +16,13 @@
 from imitation.util import networks, util
 
 ENVS = ["FrozenLake-v1", "CartPole-v1", "Pendulum-v1"]
-HARDCODED_TYPES = ["zero"]
+HARDCODED_TYPES = [
```
[the remark this answers is not in the document]
Comment (levmckinney) [reply tests/rewards/test_reward_nets.py]: Switched to a more descriptive name.

File: src/imitation/scripts/train_rl.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (yawen-d) [reply src/imitation/scripts/train_rl.py:86]: Based on our last conversations, we might want to convert it back to a warning for the purpose of this repo.

File: tests/rewards/test_reward_nets.py [hunk 17]
```diff
@@ -75,13 +72,16 @@ def _make_env_and_save_reward_net(env_name, reward_type, tmpdir):
         net = reward_nets.NormalizedRewardNet(net, networks.RunningNorm)
     elif reward_type == "RewardNet_shaped":
         net = reward_nets.ShapedRewardNet(net, _potential, discount_factor=0.99)
+    assert (
```
Comment (AdamGleave) [inline tests/rewards/test_reward_nets.py:75]: Does this function do the right thing for all reward types in the reward registry? If we add a reward to the registry, will we know to update this function?

Given this is just a helper for a *test*, I don't want to complicate things too much, but I might lean a bit towards actually just checking a hard-coded whitelist this helper was designed to work with.

**Review by AdamGleave — APPROVED** [review]
LGTM -- one minor suggestion.

File: src/imitation/scripts/train_rl.py [hunk 18]
```diff
@@ -102,6 +105,12 @@ def train_rl(
         # so normalizing it makes training more stable. Note we do *not* normalize
         # observations here; use the `NormalizeFeaturesExtractor` instead.
         venv = VecNormalize(venv, norm_obs=False, **normalize_kwargs)
+        if reward_type == "RewardNet_normalized":
+            warnings.warn(
+                "Applying normalization to already normalized reward function. \
+                Consider removing the --normalize_reward flag",
```
Comment (yawen-d) [inline src/imitation/scripts/train_rl.py:111]: Seems like `normalize_reward` isn't necessarily a flag? It's an argument in the train_rl function. Especially for Sacred experiment we usually pass in `python -m imitation.scirpts.train_rl with normalize_reward=True`. 

Suggestion: `"Consider setting normalize_reward as False"` (especially the default for normalize_reward is True in imitaion.scripts.configs.train_rl)

**Review by yawen-d — APPROVED** [review]

File: src/imitation/scripts/train_rl.py (hunk 18 above)
Comment (AdamGleave) [reply src/imitation/scripts/train_rl.py:111]: Oh yes, good catch.

Comment (levmckinney) [reply src/imitation/scripts/train_rl.py]: It is changed.

File: tests/rewards/test_reward_nets.py (hunk 17 above)
Comment (levmckinney) [reply tests/rewards/test_reward_nets.py]: I've switch over to a hard coded white list.

File: src/imitation/scripts/train_rl.py [hunk 19]
```diff
@@ -53,7 +53,9 @@ def train_rl(
 
     Args:
         total_timesteps: Number of training timesteps in `model.learn()`.
-        normalize_reward: If True, then rescale and clip reward.
+        normalize_reward: Applies normalization and clipping to the reward function by
+            keeping a running average of training rewards. Note: this is not
+            recommended if using a learned reward that is already normalized.
```
[the remark this answers is not in the document]
Comment (levmckinney) [reply src/imitation/scripts/train_rl.py:59]: I've softened the language as you suggested.

File: src/imitation/scripts/train_rl.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (levmckinney) [reply src/imitation/scripts/train_rl.py]: I converted it back to a warning.
